core/src/main/java/ai/timefold/solver/core/impl/score/director/ValueRangeState.java

```diff
@@ -0,0 +1,427 @@
+package ai.timefold.solver.core.impl.score.director;
+
+import java.util.BitSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Spliterator;
+import java.util.Spliterators;
+import java.util.function.BiFunction;
+import java.util.stream.StreamSupport;
+
+import ai.timefold.solver.core.api.domain.valuerange.CountableValueRange;
+import ai.timefold.solver.core.impl.domain.valuerange.buildin.bigdecimal.BigDecimalValueRange;
+import ai.timefold.solver.core.impl.domain.valuerange.buildin.composite.NullAllowingCountableValueRange;
+import ai.timefold.solver.core.impl.domain.valuerange.buildin.primdouble.DoubleValueRange;
+import ai.timefold.solver.core.impl.domain.valuerange.descriptor.ValueRangeDescriptor;
+import ai.timefold.solver.core.impl.domain.valuerange.sort.SelectionSorterAdapter;
+import ai.timefold.solver.core.impl.domain.valuerange.sort.SortableValueRange;
+import ai.timefold.solver.core.impl.domain.variable.descriptor.GenuineVariableDescriptor;
+import ai.timefold.solver.core.impl.heuristic.selector.common.ReachableValues;
+import ai.timefold.solver.core.impl.heuristic.selector.common.ReachableValues.ReachableValuesIndex;
+import ai.timefold.solver.core.impl.heuristic.selector.common.decorator.SelectionSorter;
+import ai.timefold.solver.core.impl.util.CollectionUtils;
+import ai.timefold.solver.core.impl.util.MutableInt;
+
+import org.jspecify.annotations.NullMarked;
+import org.jspecify.annotations.Nullable;
+
+@NullMarked
+final class ValueRangeState<Solution_, T> {
+
+    private final ValueRangeDescriptor<Solution_> valueRangeDescriptor;
+    private final Solution_ cachedWorkingSolution;
+
+    // Solution related fields
+    private @Nullable ValueRangeItem<Solution_, CountableValueRange<T>> fromSolutionItem;
+    private @Nullable Map<Object, Integer> fromSolutionValueIndexMap;
+
+    // Entity related fields
+    private @Nullable Map<Object, ValueRangeItem<Solution_, CountableValueRange<T>>> fromEntityMap;
+    private @Nullable Map<BitSetItem, Object> fromEntityBitSet;
```

Comment:
See all this use of `Object`? The code would be so much easier to understand if we used generics instead.

If it's not possible on the variable level, then at least on the method level.